src/Compiler/Microsoft.CodeAnalysis.Razor.Compiler/src/Language/RazorCSharpDocument.cs

```diff
@@ -1,81 +1,38 @@
 ﻿// Licensed to the .NET Foundation under one or more agreements.
 // The .NET Foundation licenses this file to you under the MIT license.
 
-#nullable disable
-
-using System;
-using System.Collections.Generic;
 using System.Collections.Immutable;
-using System.Linq;
 using Microsoft.AspNetCore.Razor.Language.CodeGeneration;
 
 namespace Microsoft.AspNetCore.Razor.Language;
 
-public abstract class RazorCSharpDocument : IRazorGeneratedDocument
+public sealed class RazorCSharpDocument : IRazorGeneratedDocument
 {
-    public abstract string GeneratedCode { get; }
-
-    public abstract ImmutableArray<SourceMapping> SourceMappings { get; }
-
-    public abstract IReadOnlyList<RazorDiagnostic> Diagnostics { get; }
-
-    public abstract RazorCodeGenerationOptions Options { get; }
-
-    public abstract RazorCodeDocument CodeDocument { get; }
-
-    internal virtual IReadOnlyList<LinePragma> LinePragmas { get; }
-
-    [Obsolete("For backwards compatibility only. Use the overload that takes a RazorCodeDocument")]
-    public static RazorCSharpDocument Create(string generatedCode, RazorCodeGenerationOptions options, IEnumerable<RazorDiagnostic> diagnostics)
-        => Create(codeDocument: null, generatedCode, options, diagnostics);
-
-    [Obsolete("For backwards compatibility only. Use the overload that takes a RazorCodeDocument")]
-    public static RazorCSharpDocument Create(string generatedCode, RazorCodeGenerationOptions options, IEnumerable<RazorDiagnostic> diagnostics, ImmutableArray<SourceMapping> sourceMappings, IEnumerable<LinePragma> linePragmas)
-        => Create(codeDocument: null, generatedCode, options, diagnostics, sourceMappings, linePragmas);
-
-    public static RazorCSharpDocument Create(RazorCodeDocument codeDocument, string generatedCode, RazorCodeGenerationOptions options, IEnumerable<RazorDiagnostic> diagnostics)
-    {
-        if (generatedCode == null)
-        {
-            throw new ArgumentNullException(nameof(generatedCode));
-        }
-
-        if (options == null)
-        {
-            throw new ArgumentNullException(nameof(options));
-        }
-
-        if (diagnostics == null)
-        {
-            throw new ArgumentNullException(nameof(diagnostics));
-        }
-
-        return new DefaultRazorCSharpDocument(codeDocument, generatedCode, options, diagnostics.ToArray(), sourceMappings: ImmutableArray<SourceMapping>.Empty, linePragmas: null);
-    }
-
-    public static RazorCSharpDocument Create(
+    public RazorCodeDocument CodeDocument { get; }
+    public string GeneratedCode { get; }
+    public RazorCodeGenerationOptions Options { get; }
+    public ImmutableArray<RazorDiagnostic> Diagnostics { get; }
```

Comment:
I may have missed it, why are we switching from `IReadOnlyList<>` to `ImmutableArray<>`?

Reply:
Because `IReadOnlyLsit<T>` doesn't have a struct enumerator. In fact, if the underlying type of an `IReadOnlyList<T>` _does_ have a struct enumerator (such as `List<T>`), the struct enumerator will be boxed by `IReadOnlyList<T>.GetEnumerator()`.